src/plasma/eviction_policy.cc

```diff
@@ -43,114 +43,71 @@ class LRUCache {
   }
 };
 
-/** The part of the Plasma state that is maintained by the eviction policy. */
-struct EvictionState {
-  /** The amount of memory (in bytes) currently being used. */
-  int64_t memory_used;
-  /** Datastructure for the LRU cache. */
-  LRUCache cache;
-};
-
-EvictionState *EvictionState_init() {
-  EvictionState *s = new EvictionState();
-  s->memory_used = 0;
-  return s;
-}
+EvictionPolicy::EvictionPolicy(PlasmaStoreInfo *store_info)
+    : memory_used_(0), store_info_(store_info), cache_(new LRUCache()) {}
 
-void EvictionState_free(EvictionState *s) {
-  delete s;
+EvictionPolicy::~EvictionPolicy() {
+  delete cache_;
 }
 
-int64_t EvictionState_choose_objects_to_evict(
-    EvictionState *eviction_state,
-    PlasmaStoreInfo *plasma_store_info,
+int64_t EvictionPolicy::choose_objects_to_evict(
     int64_t num_bytes_required,
-    int64_t *num_objects_to_evict,
-    ObjectID **objects_to_evict) {
-  std::vector<ObjectID> objs_to_evict;
-  int64_t bytes_evicted = eviction_state->cache.choose_objects_to_evict(
-      num_bytes_required, objs_to_evict);
+    std::vector<ObjectID> &objects_to_evict) {
+  int64_t bytes_evicted =
+      cache_->choose_objects_to_evict(num_bytes_required, objects_to_evict);
   /* Update the LRU cache. */
-  for (auto &object_id : objs_to_evict) {
-    eviction_state->cache.remove(object_id);
-  }
-  /* Construct the return values. */
-  *num_objects_to_evict = objs_to_evict.size();
-  if (objs_to_evict.size() == 0) {
-    *objects_to_evict = NULL;
-  } else {
-    int64_t result_size = objs_to_evict.size() * sizeof(ObjectID);
-    *objects_to_evict = (ObjectID *) malloc(result_size);
-    memcpy(*objects_to_evict, objs_to_evict.data(), result_size);
+  for (auto &object_id : objects_to_evict) {
+    cache_->remove(object_id);
   }
   /* Update the number of bytes used. */
-  eviction_state->memory_used -= bytes_evicted;
+  memory_used_ -= bytes_evicted;
   return bytes_evicted;
 }
 
-void EvictionState_object_created(EvictionState *eviction_state,
-                                  PlasmaStoreInfo *plasma_store_info,
-                                  ObjectID object_id) {
-  ObjectTableEntry *entry = plasma_store_info->objects[object_id];
-  eviction_state->cache.add(object_id,
-                            entry->info.data_size + entry->info.metadata_size);
+void EvictionPolicy::object_created(ObjectID object_id) {
+  auto entry = store_info_->objects[object_id].get();
```

Comment:
see this: http://stackoverflow.com/questions/17536731/shared-ptr-is-to-weak-ptr-as-unique-ptr-is-to-what

In practice people seem to use plain pointers as weak pointers for std::unique_ptr, so that's what we are doing here.